README.md
# What is RUIC?
RUIC is a Ruby API for reading, analyzing, and manipulating application assets created by [NVIDIA DRIVE™ Design][1]. Among other things, it allows you to:

* See if an application is missing any assets (e.g. images or meshes) and what parts of the application are looking for those.
* See if there are any files in the application folder that you can delete (e.g. images or materials that are no longer being used).
* Read and modify the attributes of elements on different slides.
* Batch change attributes (e.g. change all usage of one font or color to another).
* Procedurally generate many models with automated placement in your scene.

_Some of the features above are planned, but not yet implemented; see Known Limitations below._


# Documentation

In addition to the examples in this file full API documentation is available at:
http://www.rubydoc.info/gems/RUIC/



# Installing RUIC
RUIC can be installed via RubyGems (part of Ruby) via the command:

    gem install ruic   # May need `sudo gem install ruic` depending on your setup

Although RUIC is a pure-Ruby library, it relies on [Nokogiri][2] for all the XML processing and manipulation. Installing RUIC will also automatically install Nokogiri, which may require some compilation.



# Using the RUIC DSL

RUIC scripts are pure Ruby with a few convenience methods added. You run them via the `ruic` command-line script, e.g.

    ruic myscript.ruic  # or .rb extension, for syntax highlighting while editing


## Creating and Accessing Applications
RUIC scripts must start with `uia` commands to load an application and all its assets.
After this you can access the application as `app`:

```ruby
uia '../MyApp.uia'      # Relative to the ruic script file, or absolute

show app.file           #=> /var/projects/MyApp/main/MyApp.uia
show app.filename       #=> MyApp.uia

show app.assets.count   #=> 7
# You can ask for app.behaviors, app.presentations, app.statemachines, and app.renderplugins
# for arrays of specific asset types
```

_The `show` command prints the result; it is simply a nicer alias for `puts`._

If you need to load multiple applications in the same script, subsequent `uia` commands will create
`app2`, `app3`, etc. for you to use.

```ruby
uia '../MyApp.uia'       # Available as 'app'
uia '../../v1/MyApp.uia' # Available as 'app2'
```


## Working with Presentations

```ruby
uia '../MyApp.uia'

main = app.main_presentation   # The presentation displayed as the main presentation (regardless of id)
sub  = app['#nav']             # You can ask for an asset based on the id in the .uia...
sub  = app['Navigation.uip']   # or based on the path to the file (relative to the .uia)


car = sub/"Scene.Vehicle.Group.Car"      # Find elements in a presentation by presentation path…
car = app/"nav:Scene.Vehicle.Group.Car"  # …or absolute application path

show car.name #=> Car
show car.type #=> Model                  # Scene, Layer, Camera, Light, Group, Model, Material,
                                         # Image, Behavior, Effect, ReferencedMaterial, Text,
                                         # RenderPlugin, Component, (custom materials)

show car.component?  #=> false           # Ask if an element is a component
assert car.component==sub.scene          # Ask for the owning component; may be the scene
```

## Finding Many Assets

```ruby
uia 'MyApp.uia'
main = app.main_presentation

every_asset   = main.find                                  # Array of matching assets
master_assets = main.find _master:true                     # Test for master/nonmaster
models        = main.find _type:'Model'                    # …or based on type
slide2_assets = main.find _slide:2                         # …or presence on slide
rectangles    = main.find sourcepath:'#Rectangle'          # …or attribute values
gamecovers    = main.find name:'Game Cover'                # …including the name

# Combine tests to get more specific
master_models = main.find _type:'Model', _master:true
slide2_rects  = main.find _type:'Model', _slide:2, sourcepath:'#Rectangle'
nonmaster_s2  = main.find _slide:2, _master:false
red_materials = main.find _type:'Material', diffuse:[1,0,0]

# You can match values more loosely
pistons       = main.find name:/^Piston/                   # Regex for batch finding
bottom_row    = main.find position:[nil,-200,nil]          # nil for wildcards in vectors

# Restrict the search to a sub-tree
group        = main/"Scene.Layer.Group"
group_models = group.find _type:'Model'                    # Orig asset is never included
group_models = main.find _under:group, _type:'Model'       # Or, use `_under` to limit

# Iterate the results as they are found
main.find _type:'Model', name:/^Piston/ do |model, index|  # Using the index is optional
	show "Model #{index} is named #{model.name}"
end
```

Notes:

* `nil` inside an array is a "wildcard" value, allowing you to test only specific values
* Numbers (both in vectors/colors/rotations and float/long values) must only be within `0.001` to match.
  * _For example, `attr:{diffuse:[1,0,0]}` will match a color with `diffuse=".9997 0.0003 0"`_
* Results of `find` are always in scene-graph order.


## Working with References

```ruby
uia 'MyApp.uia'
mat1 = app/"main:Scene.Layer.Sphere.Material" # A standard NDD Material
mat2 = app/"main:Scene.Layer.Cube.Material"   # A standard NDD Material
p mat2.type                                   #=> "Material"
ref = mat2.replace_with_referenced_material   # A very specific method :)
p ref.properties['referencedmaterial'].type   #=> "ObjectRef"
p ref['referencedmaterial',0].object          #=> nil
p ref['referencedmaterial',0].type            #=> :absolute
ref['referencedmaterial',0].object = mat1     #=> Sets an absolute reference
ref['referencedmaterial',0].type = :path      #=> Use a relative path instead

# Alternatively, you can omit the .object when setting the reference:
# ref['referencedmaterial',0] = mat1

mat3 = ref['referencedmaterial',1].object     #=> Get the asset pointed to
assert mat1 == mat3                           #=> They are the same! It worked!

app.save_all!                                 #=> Write presentations in place
```


## Locating MetaData.xml
RUIC needs access to a NDD `MetaData.xml` file to understand the properties in the various XML files.
By default RUIC will look in the location specified by `RUIC::DEFAULTMETADATA`, e.g.
`C:\Program Files (x86)\NVIDIA Corporation\DRIVE Design 8.5\res\DataModelMetadata\en-us\MetaData.xml`

If this file is in another location, you can tell the script where to find it either:

* on the command line: `ruic -m path/to/MetaData.xml myscript.ruic`
* in your ruic script: `metadata 'path/to/MetaData.xml' # before any 'app' commands`


# Interactive RUIC
In addition to executing a standalone script, RUIC also has a REPL (like IRB) that allows you to
interactively execute and test changes to your application/presentations before saving.
There are two ways to enter interactive mode:

* If you invoke the `ruic` binary with a `.uia` file as the argument the interpreter will load
  the application and enter the REPL:


HISTORY
## v0.7.0

### Backwards Incompatible Changes

* Updated to reflect marketing change from "UI Composer" to "NVIDIA DRIVE™ Design".
  * The `UIC` module is now named `NDD`.


## v0.6.1

### New/Changed Features

* Support for the new `PathBuffer` property types (for paths that reference a `.path` on disk).
* Unsupported/invalid custom property types fall back to String (with a warning).

### Bug Fixes
* Render plugins are now lightly tested and supported.


## v0.6.0

### Backwards Incompatible Changes

* Changed `FileBacked#resolve_file_path` to `FileBacked#absolute_path`, and added `FileBacked#relative_path`.
* Removed `.errors?` and `.errors` for application and assets.

### New/Changed Features

* State machines properly report on visual actions.
  * _Only `<set-attribute>` properly implemented so far._
* Added much improved `Presentation#referenced_files`, `Application#missing_files`, `Application#unused_files`.
* Tweaked REPL output.
  * `nil` results no longer print `#=> nil` in the REPL.
  * `show` prefixes the result with `#=>`.
  * Multi-line results are now wrapped, with each line prefixed with `#=>`.
* Added beginnings of support for effects and renderplugins.
* Presentations do not cause a runtime error if a custom class (e.g. behavior, effect, custom material) cannot be found.

### Bug Fixes
* Fix bug that caused some properties to report their type as `nil` or `float` instead of `Float`.
* Presentations are slightly more capable if they are created without an existing file.


## v0.5.0 - 2014-Nov-26

* Fix bug that prevented loading an application from the command line in another directory.
* Compatibility: `FileBacked#path_to` is now `FileBacked#resolve_file_path` to prevent conflict with `Presentation#path_to`.
* Make `inspect` representation of assets more useful.
* Add convenience methods for collections of assets by type on presentations:
  `.layers`, `.groups`, `.cameras`, `.lights`, `.components`, `.models`, `.materials`,
  `.images`, `.behaviors`, `.aliases`, `.paths`, `.anchor_points`, `.texts`
* Fix bug that prevented behavior instances from having proper `name` or `sourcepath` attribute values.
* Added `.hierarchy` to presentations and assets for printing the scene graph.


## v0.4.5 - 2014-Nov-19

* Added a large amount of documentation.
* Minor code refactoring:
  * Most (internal) Presentation methods that took a scene graph element as the argument now take an asset.
  * All `FileBacked` assets now have `save!` and `save_as`.


## v0.4.1 - 2014-Nov-16

* Update `MetaData.xml` for tests to newest version.
* Source paths for images/textures normalize the path (forward slashes and no leading `./`).
* Added `app.main` as a simpler alias for `app.main_presentation`.
* Add support for paths and anchor points, including `mypath.anchors` as an array of anchor points.
* Assets present in MetaData and presentation, but not yet hand-entered into the hierarchy, now work and default to inherit from `Asset`.


## v0.4.0 - 2014-Nov-11
* Switch attribute filtering again; now all attributes are inline in the `find()` hash, and the four special keys are prefixed with an underscore: `_type`, `_slide`, `_master`, and `_under`.


## v0.3.0 - 2014-Nov-10
* Switch attribute filtering to use `attr:{ … }` instead of `attributes:{ … }`
* Attribute matching now requires that a requested attribute be present, or else the asset matching fails.
  * _For example, `main.find attr:{ diffusecolor:[nil,nil,nil] }` will now only find assets with a `diffusecolor` attribute._


## v0.2.5 - 2014-Nov-10

gui/window_ui.rb
=begin
** Form generated from reading ui file 'window.ui'
**
** Created: Tue Oct 7 14:55:11 2014
**      by: Qt User Interface Compiler version 4.8.6
**
** WARNING! All changes made in this file will be lost when recompiling ui file!
=end

class Ui_MainWin
    attr_reader :menuFileOpen
    attr_reader :menuEditUndo
    attr_reader :menuEditRedo
    attr_reader :menuEditCopy
    attr_reader :actionSave_All
    attr_reader :actionQuit
    attr_reader :actionDelete_Unused_Assets
    attr_reader :centralwidget
    attr_reader :verticalLayout_4
    attr_reader :mainSplitter
    attr_reader :topRowSplitter
    attr_reader :layoutWidget
    attr_reader :elementsBox
    attr_reader :elementsHeader
    attr_reader :elementsSplitter
    attr_reader :elements
    attr_reader :layoutWidget1
    attr_reader :verticalLayout
    attr_reader :slideTabs
    attr_reader :tab_3
    attr_reader :horizontalLayout
    attr_reader :tab_4
    attr_reader :tab_5
    attr_reader :inspector
    attr_reader :layoutWidget2
    attr_reader :valuesBox
    attr_reader :valuesHeaderBox
    attr_reader :valuesHeader
    attr_reader :hspace
    attr_reader :comboBox_2
    attr_reader :comboBox_3
    attr_reader :comboBox_4
    attr_reader :tableView
    attr_reader :layoutWidget3
    attr_reader :assetsBox
    attr_reader :assetsHeaderBox
    attr_reader :assetsHeader
    attr_reader :hspace_2
    attr_reader :comboBox
    attr_reader :assetView
    attr_reader :layoutWidget4
    attr_reader :consoleBox
    attr_reader :consoleLabel
    attr_reader :console
    attr_reader :menubar
    attr_reader :menuFile
    attr_reader :menuEdit
    attr_reader :menuWindow
    attr_reader :statusbar

    def setupUi(mainWin)
    if mainWin.objectName.nil?
        mainWin.objectName = "mainWin"
    end
    mainWin.resize(1150, 656)
    mainWin.styleSheet = "QPlainTextEdit, QTableView, QTreeView { border:1px solid #898c95 }"
    mainWin.unifiedTitleAndToolBarOnMac = false
    @menuFileOpen = Qt::Action.new(mainWin)
    @menuFileOpen.objectName = "menuFileOpen"
    icon = Qt::Icon.new
    icon.addPixmap(Qt::Pixmap.new(":/resources/images/folder-horizontal-open.png"), Qt::Icon::Normal, Qt::Icon::Off)
    @menuFileOpen.icon = icon
    @menuFileOpen.shortcutContext = Qt::ApplicationShortcut
    @menuEditUndo = Qt::Action.new(mainWin)
    @menuEditUndo.objectName = "menuEditUndo"
    @menuEditUndo.enabled = false
    icon1 = Qt::Icon.new
    icon1.addPixmap(Qt::Pixmap.new(), Qt::Icon::Normal, Qt::Icon::Off)
    @menuEditUndo.icon = icon1
    @menuEditRedo = Qt::Action.new(mainWin)
    @menuEditRedo.objectName = "menuEditRedo"
    @menuEditRedo.enabled = false
    @menuEditCopy = Qt::Action.new(mainWin)
    @menuEditCopy.objectName = "menuEditCopy"
    @menuEditCopy.enabled = false
    @actionSave_All = Qt::Action.new(mainWin)
    @actionSave_All.objectName = "actionSave_All"
    @actionSave_All.enabled = false
    icon2 = Qt::Icon.new
    icon2.addPixmap(Qt::Pixmap.new(":/resources/images/disks-black.png"), Qt::Icon::Normal, Qt::Icon::Off)
    @actionSave_All.icon = icon2
    @actionQuit = Qt::Action.new(mainWin)
    @actionQuit.objectName = "actionQuit"
    @actionDelete_Unused_Assets = Qt::Action.new(mainWin)
    @actionDelete_Unused_Assets.objectName = "actionDelete_Unused_Assets"
    icon3 = Qt::Icon.new
    icon3.addPixmap(Qt::Pixmap.new(":/resources/images/cross.png"), Qt::Icon::Normal, Qt::Icon::Off)
    @actionDelete_Unused_Assets.icon = icon3
    @centralwidget = Qt::Widget.new(mainWin)
    @centralwidget.objectName = "centralwidget"
    @sizePolicy = Qt::SizePolicy.new(Qt::SizePolicy::Preferred, Qt::SizePolicy::Minimum)
    @sizePolicy.setHorizontalStretch(0)
    @sizePolicy.setVerticalStretch(0)
    @sizePolicy.heightForWidth = @centralwidget.sizePolicy.hasHeightForWidth
    @centralwidget.sizePolicy = @sizePolicy
    @centralwidget.autoFillBackground = true
    @verticalLayout_4 = Qt::VBoxLayout.new(@centralwidget)
    @verticalLayout_4.objectName = "verticalLayout_4"
    @mainSplitter = Qt::Splitter.new(@centralwidget)
    @mainSplitter.objectName = "mainSplitter"
    @mainSplitter.lineWidth = 0
    @mainSplitter.orientation = Qt::Vertical
    @mainSplitter.handleWidth = 9
    @topRowSplitter = Qt::Splitter.new(@mainSplitter)
    @topRowSplitter.objectName = "topRowSplitter"
    @sizePolicy1 = Qt::SizePolicy.new(Qt::SizePolicy::Expanding, Qt::SizePolicy::Preferred)
    @sizePolicy1.setHorizontalStretch(0)
    @sizePolicy1.setVerticalStretch(1)
    @sizePolicy1.heightForWidth = @topRowSplitter.sizePolicy.hasHeightForWidth
    @topRowSplitter.sizePolicy = @sizePolicy1
    @topRowSplitter.orientation = Qt::Horizontal
    @topRowSplitter.handleWidth = 9
    @topRowSplitter.childrenCollapsible = false
    @layoutWidget = Qt::Widget.new(@topRowSplitter)
    @layoutWidget.objectName = "layoutWidget"
    @elementsBox = Qt::VBoxLayout.new(@layoutWidget)
    @elementsBox.spacing = 0
    @elementsBox.objectName = "elementsBox"
    @elementsBox.setContentsMargins(0, 0, 0, 0)
    @elementsHeader = Qt::Label.new(@layoutWidget)
    @elementsHeader.objectName = "elementsHeader"
    @sizePolicy2 = Qt::SizePolicy.new(Qt::SizePolicy::Preferred, Qt::SizePolicy::Fixed)
    @sizePolicy2.setHorizontalStretch(0)
    @sizePolicy2.setVerticalStretch(0)
    @sizePolicy2.heightForWidth = @elementsHeader.sizePolicy.hasHeightForWidth
    @elementsHeader.sizePolicy = @sizePolicy2
    @elementsHeader.minimumSize = Qt::Size.new(61, 16)
    @elementsHeader.maximumSize = Qt::Size.new(16777215, 16)
    @font = Qt::Font.new
    @font.bold = true
    @font.weight = 75
    @elementsHeader.font = @font

    @elementsBox.addWidget(@elementsHeader)

    @elementsSplitter = Qt::Splitter.new(@layoutWidget)
    @elementsSplitter.objectName = "elementsSplitter"
    @sizePolicy3 = Qt::SizePolicy.new(Qt::SizePolicy::Expanding, Qt::SizePolicy::Expanding)
    @sizePolicy3.setHorizontalStretch(0)
    @sizePolicy3.setVerticalStretch(0)
    @sizePolicy3.heightForWidth = @elementsSplitter.sizePolicy.hasHeightForWidth
    @elementsSplitter.sizePolicy = @sizePolicy3
    @elementsSplitter.minimumSize = Qt::Size.new(361, 0)
    @elementsSplitter.maximumSize = Qt::Size.new(16777215, 16777215)
    @elementsSplitter.orientation = Qt::Horizontal
    @elementsSplitter.handleWidth = 9
    @elementsSplitter.childrenCollapsible = false
    @elements = Qt::TreeView.new(@elementsSplitter)
    @elements.objectName = "elements"
    @elements.minimumSize = Qt::Size.new(0, 60)
    @elements.styleSheet = ""
    @elements.lineWidth = 1
    @elements.horizontalScrollBarPolicy = Qt::ScrollBarAlwaysOff
    @elements.uniformRowHeights = true
    @elements.headerHidden = false
    @elementsSplitter.addWidget(@elements)
    @layoutWidget1 = Qt::Widget.new(@elementsSplitter)
    @layoutWidget1.objectName = "layoutWidget1"
    @verticalLayout = Qt::VBoxLayout.new(@layoutWidget1)
    @verticalLayout.spacing = 0
    @verticalLayout.objectName = "verticalLayout"
    @verticalLayout.setContentsMargins(0, 0, 0, 0)
    @slideTabs = Qt::TabWidget.new(@layoutWidget1)
    @slideTabs.objectName = "slideTabs"
    @sizePolicy4 = Qt::SizePolicy.new(Qt::SizePolicy::Expanding, Qt::SizePolicy::Fixed)
    @sizePolicy4.setHorizontalStretch(0)
    @sizePolicy4.setVerticalStretch(0)
    @sizePolicy4.heightForWidth = @slideTabs.sizePolicy.hasHeightForWidth
    @slideTabs.sizePolicy = @sizePolicy4
    @slideTabs.maximumSize = Qt::Size.new(16777215, 21)
    @slideTabs.iconSize = Qt::Size.new(32, 32)
    @tab_3 = Qt::Widget.new()
    @tab_3.objectName = "tab_3"
    @tab_3.maximumSize = Qt::Size.new(16777215, 0)
    @horizontalLayout = Qt::HBoxLayout.new(@tab_3)
    @horizontalLayout.objectName = "horizontalLayout"
    @slideTabs.addTab(@tab_3, Qt::Application.translate("MainWin", "0: Master", nil, Qt::Application::UnicodeUTF8))
    @tab_4 = Qt::Widget.new()
    @tab_4.objectName = "tab_4"
    @tab_4.maximumSize = Qt::Size.new(16777215, 0)
    @slideTabs.addTab(@tab_4, Qt::Application.translate("MainWin", "1: Slide1", nil, Qt::Application::UnicodeUTF8))
    @tab_5 = Qt::Widget.new()
    @tab_5.objectName = "tab_5"
    @slideTabs.addTab(@tab_5, Qt::Application.translate("MainWin", "2: Slide2", nil, Qt::Application::UnicodeUTF8))

    @verticalLayout.addWidget(@slideTabs)

    @inspector = Qt::TableView.new(@layoutWidget1)
    @inspector.objectName = "inspector"
    @inspector.minimumSize = Qt::Size.new(0, 60)
    @inspector.styleSheet = "#inspector { border-top:none; margin-right:2px /*hack to line up with border on tabs */ }"
    @inspector.horizontalScrollBarPolicy = Qt::ScrollBarAlwaysOff
    @inspector.showGrid = true
    @inspector.wordWrap = false

    @verticalLayout.addWidget(@inspector)

    @elementsSplitter.addWidget(@layoutWidget1)

    @elementsBox.addWidget(@elementsSplitter)

    @topRowSplitter.addWidget(@layoutWidget)
    @layoutWidget2 = Qt::Widget.new(@topRowSplitter)
    @layoutWidget2.objectName = "layoutWidget2"
    @valuesBox = Qt::VBoxLayout.new(@layoutWidget2)
    @valuesBox.spacing = 0
    @valuesBox.objectName = "valuesBox"
    @valuesBox.setContentsMargins(0, 0, 0, 0)
    @valuesHeaderBox = Qt::HBoxLayout.new()
    @valuesHeaderBox.spacing = 2
    @valuesHeaderBox.objectName = "valuesHeaderBox"
    @valuesHeaderBox.setContentsMargins(-1, -1, -1, 1)
    @valuesHeader = Qt::Label.new(@layoutWidget2)
    @valuesHeader.objectName = "valuesHeader"
    @sizePolicy2.heightForWidth = @valuesHeader.sizePolicy.hasHeightForWidth
    @valuesHeader.sizePolicy = @sizePolicy2
    @valuesHeader.minimumSize = Qt::Size.new(45, 0)
    @valuesHeader.maximumSize = Qt::Size.new(16777215, 16)
    @valuesHeader.font = @font

    @valuesHeaderBox.addWidget(@valuesHeader)

    @hspace = Qt::SpacerItem.new(48, 13, Qt::SizePolicy::Expanding, Qt::SizePolicy::Minimum)

    @valuesHeaderBox.addItem(@hspace)

    @comboBox_2 = Qt::ComboBox.new(@layoutWidget2)
    @comboBox_2.objectName = "comboBox_2"
    @comboBox_2.minimumSize = Qt::Size.new(76, 0)
    @comboBox_2.maximumSize = Qt::Size.new(87, 16)
    @comboBox_2.maxVisibleItems = 15

    @valuesHeaderBox.addWidget(@comboBox_2)

    @comboBox_3 = Qt::ComboBox.new(@layoutWidget2)
    @comboBox_3.objectName = "comboBox_3"
    @comboBox_3.minimumSize = Qt::Size.new(69, 0)
    @comboBox_3.maximumSize = Qt::Size.new(69, 16)

    @valuesHeaderBox.addWidget(@comboBox_3)

    @comboBox_4 = Qt::ComboBox.new(@layoutWidget2)
    @comboBox_4.objectName = "comboBox_4"
    @sizePolicy5 = Qt::SizePolicy.new(Qt::SizePolicy::Minimum, Qt::SizePolicy::Fixed)
    @sizePolicy5.setHorizontalStretch(0)
    @sizePolicy5.setVerticalStretch(0)
    @sizePolicy5.heightForWidth = @comboBox_4.sizePolicy.hasHeightForWidth
    @comboBox_4.sizePolicy = @sizePolicy5
    @comboBox_4.minimumSize = Qt::Size.new(80, 0)
    @comboBox_4.maximumSize = Qt::Size.new(16777215, 16)

    @valuesHeaderBox.addWidget(@comboBox_4)


    @valuesBox.addLayout(@valuesHeaderBox)

    @tableView = Qt::TableView.new(@layoutWidget2)
    @tableView.objectName = "tableView"
    @tableView.minimumSize = Qt::Size.new(0, 60)

    @valuesBox.addWidget(@tableView)

    @topRowSplitter.addWidget(@layoutWidget2)
    @layoutWidget3 = Qt::Widget.new(@topRowSplitter)
    @layoutWidget3.objectName = "layoutWidget3"
    @assetsBox = Qt::VBoxLayout.new(@layoutWidget3)
    @assetsBox.spacing = 0
    @assetsBox.objectName = "assetsBox"
    @assetsBox.setContentsMargins(0, 0, 0, 0)
    @assetsHeaderBox = Qt::HBoxLayout.new()
    @assetsHeaderBox.objectName = "assetsHeaderBox"
    @assetsHeaderBox.setContentsMargins(-1, -1, -1, 1)
    @assetsHeader = Qt::Label.new(@layoutWidget3)
    @assetsHeader.objectName = "assetsHeader"
    @sizePolicy2.heightForWidth = @assetsHeader.sizePolicy.hasHeightForWidth
    @assetsHeader.sizePolicy = @sizePolicy2
    @assetsHeader.minimumSize = Qt::Size.new(46, 16)
    @assetsHeader.maximumSize = Qt::Size.new(16777215, 16)
    @assetsHeader.font = @font

    @assetsHeaderBox.addWidget(@assetsHeader)

    @hspace_2 = Qt::SpacerItem.new(40, 16, Qt::SizePolicy::Expanding, Qt::SizePolicy::Minimum)

    @assetsHeaderBox.addItem(@hspace_2)

    @comboBox = Qt::ComboBox.new(@layoutWidget3)
    @comboBox.objectName = "comboBox"
    @comboBox.minimumSize = Qt::Size.new(69, 16)
    @comboBox.maximumSize = Qt::Size.new(16777215, 16)

    @assetsHeaderBox.addWidget(@comboBox)


    @assetsBox.addLayout(@assetsHeaderBox)

    @assetView = Qt::TreeView.new(@layoutWidget3)
    @assetView.objectName = "assetView"
    @sizePolicy3.heightForWidth = @assetView.sizePolicy.hasHeightForWidth
    @assetView.sizePolicy = @sizePolicy3
    @assetView.minimumSize = Qt::Size.new(131, 101)

    @assetsBox.addWidget(@assetView)

    @topRowSplitter.addWidget(@layoutWidget3)
    @mainSplitter.addWidget(@topRowSplitter)
    @layoutWidget4 = Qt::Widget.new(@mainSplitter)
    @layoutWidget4.objectName = "layoutWidget4"
    @consoleBox = Qt::VBoxLayout.new(@layoutWidget4)
    @consoleBox.spacing = 0
    @consoleBox.objectName = "consoleBox"
    @consoleBox.setContentsMargins(0, 0, 0, 0)
    @consoleLabel = Qt::Label.new(@layoutWidget4)
    @consoleLabel.objectName = "consoleLabel"
    @sizePolicy2.heightForWidth = @consoleLabel.sizePolicy.hasHeightForWidth
    @consoleLabel.sizePolicy = @sizePolicy2
    @consoleLabel.maximumSize = Qt::Size.new(16777215, 16)
    @consoleLabel.font = @font

    @consoleBox.addWidget(@consoleLabel)

    @console = Qt::PlainTextEdit.new(@layoutWidget4)
    @console.objectName = "console"
    @console.enabled = true
    @sizePolicy6 = Qt::SizePolicy.new(Qt::SizePolicy::Ignored, Qt::SizePolicy::Minimum)
    @sizePolicy6.setHorizontalStretch(0)
    @sizePolicy6.setVerticalStretch(0)
    @sizePolicy6.heightForWidth = @console.sizePolicy.hasHeightForWidth
    @console.sizePolicy = @sizePolicy6
    @console.minimumSize = Qt::Size.new(0, 41)
    @console.undoRedoEnabled = false
    @console.readOnly = true
    @console.tabStopWidth = 2
    @console.textInteractionFlags = Qt::LinksAccessibleByMouse|Qt::TextSelectableByMouse

    @consoleBox.addWidget(@console)

    @mainSplitter.addWidget(@layoutWidget4)

    @verticalLayout_4.addWidget(@mainSplitter)

    mainWin.centralWidget = @centralwidget
    @menubar = Qt::MenuBar.new(mainWin)
    @menubar.objectName = "menubar"
    @menubar.geometry = Qt::Rect.new(0, 0, 1150, 21)
    @menuFile = Qt::Menu.new(@menubar)
    @menuFile.objectName = "menuFile"
    @menuEdit = Qt::Menu.new(@menubar)
    @menuEdit.objectName = "menuEdit"
    @menuWindow = Qt::Menu.new(@menubar)
    @menuWindow.objectName = "menuWindow"
    mainWin.setMenuBar(@menubar)
    @statusbar = Qt::StatusBar.new(mainWin)
    @statusbar.objectName = "statusbar"
    mainWin.statusBar = @statusbar

    @menubar.addAction(@menuFile.menuAction())
    @menubar.addAction(@menuEdit.menuAction())
    @menubar.addAction(@menuWindow.menuAction())
    @menuFile.addAction(@menuFileOpen)
    @menuFile.addAction(@actionSave_All)
    @menuFile.addSeparator()
    @menuFile.addAction(@actionDelete_Unused_Assets)
    @menuFile.addSeparator()
    @menuFile.addAction(@actionQuit)
    @menuEdit.addAction(@menuEditUndo)
    @menuEdit.addAction(@menuEditRedo)
    @menuEdit.addSeparator()
    @menuEdit.addAction(@menuEditCopy)

    retranslateUi(mainWin)

    @slideTabs.setCurrentIndex(0)


    Qt::MetaObject.connectSlotsByName(mainWin)
    end # setupUi

    def setup_ui(mainWin)
        setupUi(mainWin)
    end

    def retranslateUi(mainWin)
    mainWin.windowTitle = Qt::Application.translate("MainWin", "UIC Inspectamator", nil, Qt::Application::UnicodeUTF8)
    @menuFileOpen.text = Qt::Application.translate("MainWin", "Open\342\200\246", nil, Qt::Application::UnicodeUTF8)
    @menuFileOpen.shortcut = Qt::Application.translate("MainWin", "Ctrl+O", nil, Qt::Application::UnicodeUTF8)
    @menuEditUndo.text = Qt::Application.translate("MainWin", "Undo", nil, Qt::Application::UnicodeUTF8)
    @menuEditUndo.shortcut = Qt::Application.translate("MainWin", "Ctrl+Z", nil, Qt::Application::UnicodeUTF8)
    @menuEditRedo.text = Qt::Application.translate("MainWin", "Redo", nil, Qt::Application::UnicodeUTF8)
    @menuEditRedo.shortcut = Qt::Application.translate("MainWin", "Ctrl+Y", nil, Qt::Application::UnicodeUTF8)
    @menuEditCopy.text = Qt::Application.translate("MainWin", "Copy", nil, Qt::Application::UnicodeUTF8)
    @menuEditCopy.shortcut = Qt::Application.translate("MainWin", "Ctrl+C", nil, Qt::Application::UnicodeUTF8)
    @actionSave_All.text = Qt::Application.translate("MainWin", "Save All\342\200\246", nil, Qt::Application::UnicodeUTF8)
    @actionSave_All.shortcut = Qt::Application.translate("MainWin", "Ctrl+S", nil, Qt::Application::UnicodeUTF8)
    @actionQuit.text = Qt::Application.translate("MainWin", "Quit", nil, Qt::Application::UnicodeUTF8)
    @actionDelete_Unused_Assets.text = Qt::Application.translate("MainWin", "Delete Unused Assets\342\200\246", nil, Qt::Application::UnicodeUTF8)
    @elementsHeader.text = Qt::Application.translate("MainWin", "Elements", nil, Qt::Application::UnicodeUTF8)
    @slideTabs.setTabText(@slideTabs.indexOf(@tab_3), Qt::Application.translate("MainWin", "0: Master", nil, Qt::Application::UnicodeUTF8))
    @slideTabs.setTabText(@slideTabs.indexOf(@tab_4), Qt::Application.translate("MainWin", "1: Slide1", nil, Qt::Application::UnicodeUTF8))
    @slideTabs.setTabText(@slideTabs.indexOf(@tab_5), Qt::Application.translate("MainWin", "2: Slide2", nil, Qt::Application::UnicodeUTF8))
    @valuesHeader.text = Qt::Application.translate("MainWin", "Values", nil, Qt::Application::UnicodeUTF8)
    @comboBox_2.insertItems(0, [Qt::Application.translate("MainWin", "(elements)", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "scenes", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "layers", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "cameras", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "lights", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "groups", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "models", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "materials", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "text", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "components", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "behaviors", nil, Qt::Application::UnicodeUTF8)])
    @comboBox_3.insertItems(0, [Qt::Application.translate("MainWin", "(types)", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "floats", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "integers", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "colors", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "strings", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "fonts", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "fontsizes", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "vectors", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "rotations", nil, Qt::Application::UnicodeUTF8)])
    @comboBox_4.insertItems(0, [Qt::Application.translate("MainWin", "(attributes)", nil, Qt::Application::UnicodeUTF8)])
    @assetsHeader.text = Qt::Application.translate("MainWin", "Assets", nil, Qt::Application::UnicodeUTF8)
    @comboBox.insertItems(0, [Qt::Application.translate("MainWin", "All Assets", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "Used Only", nil, Qt::Application::UnicodeUTF8),
        Qt::Application.translate("MainWin", "Unused Only", nil, Qt::Application::UnicodeUTF8)])
    @consoleLabel.text = Qt::Application.translate("MainWin", "Console", nil, Qt::Application::UnicodeUTF8)
    @menuFile.title = Qt::Application.translate("MainWin", "File", nil, Qt::Application::UnicodeUTF8)
    @menuEdit.title = Qt::Application.translate("MainWin", "Edit", nil, Qt::Application::UnicodeUTF8)
    @menuWindow.title = Qt::Application.translate("MainWin", "Window", nil, Qt::Application::UnicodeUTF8)
    end # retranslateUi

    def retranslate_ui(mainWin)
        retranslateUi(mainWin)
    end

end

module Ui
    class MainWin < Ui_MainWin
    end
end  # module Ui

